### PR TITLE
Disable doublefire

### PR DESCRIPTION
**Pull request by EatThePath**
Implements the requested features in #1851 

Doublefiring can be turned off for AIs and/or players separately through AI profiles flags. Went with that instead of gamesettings largely because that seems consistent with other similar behavior options, and I can conceivably see someone wanting to vary it between missions. 

It can also be turned off on a per-weapon basis through a new weapon flag.

### CODE REVIEW

**Comment by EatThePath** [comment]
Noted a logical issue when player doublefire is off but AI doublefire is on, testing a fix momentarily.

**Comment by EatThePath** [comment]
My fix isn't working in mysterious ways, and I'm not going to figure it out tonight, so I'm closing this until I can sort it.

**Comment by EatThePath** [comment]
The previous problem case has been resolved.

File: code/ai/aicode.cpp [hunk 1]
```diff
@@ -5520,7 +5520,12 @@ void set_primary_weapon_linkage(object *objp)
 				if (!sip->flags[Ship::Info_Flags::No_primary_linking] ) {
 					shipp->flags.set(Ship::Ship_Flags::Primary_linked);
 				}
-                shipp->flags.set(Ship::Ship_Flags::Secondary_dual_fire);
+				if (The_mission.ai_profile->flags[AI::Profile_Flags::Disable_ai_secondary_doublefire] || 
+					Weapon_info[swp->secondary_bank_weapons[swp->current_secondary_bank]].wi_flags[Weapon::Info_Flags::No_doublefire]) {
+					shipp->flags.remove(Ship::Ship_Flags::Secondary_dual_fire);
+				} else {
```
[the remark this answers is not in the document]
Comment (EatThePath) [reply code/ai/aicode.cpp:5526]: I'd assumed there were other places where the flag might be set for AI ships, like a sexp. On further review that doesn't seem to be the case. I would still personally prefer to safeguard against the possibility of that in the future, and on another look at the function it feels more sane to put it a little further up where the primary link flag is unset as well, but if that's unwelcome I suppose it's not strictly needed.

File: code/io/keycontrol.cpp [hunk 2]
```diff
@@ -1888,8 +1888,9 @@ int button_function_critical(int n, net_player *p = NULL)
 
 			int firepoints = pm->missile_banks[Ships[objp->instance].weapons.current_secondary_bank].num_slots;
 
-            if (Ships[objp->instance].flags[Ship::Ship_Flags::Secondary_dual_fire] || firepoints < 2) {
-                Ships[objp->instance].flags.remove(Ship::Ship_Flags::Secondary_dual_fire);
+			if (Ships[objp->instance].flags[Ship::Ship_Flags::Secondary_dual_fire] || firepoints < 2 ||
+				The_mission.ai_profile->flags[AI::Profile_Flags::Disable_player_secondary_doublefire]) {
+				Ships[objp->instance].flags.remove(Ship::Ship_Flags::Secondary_dual_fire);
```
[the remark this answers is not in the document]
Comment (EatThePath) [reply code/io/keycontrol.cpp:1893]: To the first, my thinking was that it might be set if the player previously had it set on a save where this flag wasn't used. I'm frankly not certain if that's possible, but it seemed harmless to be sure. Strictly speaking this is safeguarded by the check at firing time, but doing it here provides more immediate player feedback on the possibility of firing two missiles. 

To the second, I don't understand how that could be the case. I'm also not sure it's especially undesirable, again giving the player more immediate information.

On the gripping hand though this doesn't currently handle the case of the individual weapon flag, so even if I prevail on the first two points it still needs work, though different work.

File: code/ship/ship.cpp [hunk 3]
```diff
@@ -11921,7 +11922,16 @@ int ship_fire_secondary( object *obj, int allow_swarm )
...
+		// Checking for the secondary doublefire disabling here in case it slipped, such as being set to double fire in
+		// a previous session If doublefire is set and((player double fire is disabled and this is the player) or 
+		/// (this weapon's doublefire is disabled))
+		if (shipp->flags[Ship_Flags::Secondary_dual_fire] &&
+			((The_mission.ai_profile->flags[AI::Profile_Flags::Disable_player_secondary_doublefire] &&  obj == Player_obj) ||
+			 Weapon_info[swp->secondary_bank_weapons[swp->current_secondary_bank]].wi_flags[Weapon::Info_Flags::No_doublefire])) {
+			shipp->flags.remove(Ship_Flags::Secondary_dual_fire);
+		}
+
+		if (shipp->flags[Ship_Flags::Secondary_dual_fire] && num_slots > 1) {
```
[the remark this answers is not in the document]
Comment (EatThePath) [reply code/ship/ship.cpp:11934]: It's less about not trusting my code elsewhere as it is not being certain of all the other code in the game.

I have now tested the one corner case I had in mind. If you go from a version of the game/a mod without the flag to one with, you can keep doublefire set from before. It's a very niche case yes, but it is a possible one.  

Also possible in the case of the weapon flag version is having doublefire set on bank 0 and then swapping to a no-doublefire weapon in bank 1. You could have the check in keycontrol when switching secondary banks, but by doing it here you allow the player to cycle past a no-doublefire bank to another bank without losing the doublefire setting. That does come at the cost of not letting the player see immediately that they've switched to a no-doublefire, but without adding some new state it's a pick your poison situation and this is the option I personally think is the better player experience.

File: code/ai/aicode.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Goober5000) [reply code/ai/aicode.cpp:5526]: Then use an Assertion, as I mentioned in one of the other sections.

File: code/io/keycontrol.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Goober5000) [reply code/io/keycontrol.cpp:1893]: Again, use an Assertion.

"every time the mission began" is not strictly true; let me rephrase.  This change would take a block of code that's not supposed to run and not expected to run, due to the lack of secondary linking.  But when you check the ai_profiles flag it *does* run and then immediately reverts what it did.  This is bad logic.

File: code/ship/ship.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Goober5000) [reply code/ship/ship.cpp:11934]: Again, if you're not certain of the other code in the game then use an Assertion.  An Assertion is meant to guarantee that the code contract behaves as it's expected to.

Why would you ever go from two versions of the mod in the *same* mission?  Ship flags are supposed to be transient, only existing during the lifetime of one mission.  In order to switch mods you need to exit not only the mission but also the game.

I think it's a better player experience to remove the doublefire flag as soon as the player switches to a no-doublefire bank.  Hiding important feedback from the player is a bad thing.

[the remark this answers is not in the document]
Comment (EatThePath) [reply code/ship/ship.cpp:11934]: I only have the free mental bandwidth to address this point right now, I'll have to circle back for the rest later.

Maybe ship flags are supposed to be transient, but this one is a bit sticky. Based on my testing, there are many ways that a user could start the mission with an invalid doublefire state.

- While developing a mod or mission, after first disallowing doublefire
- When updating to a new version of a mod or mission that didn't disallow doublefire but now does
- When doublefire was allowed in the last played mission but isn't in the next.
- When replaying the same mission but switching from a doublefire allowed missile loadout to a doublefire disallowed loadout.

I haven't rooted through the code to figure out why this state persists between missions yet but my experimental testing indicates it does, at least in tech room missions. I would expect campaign is the same but haven't tested it.

Is an assertion still the right way to handle all these cases? Or perhaps is this persistence something that needs to be hunted down and destroyed in the first place? I'm not sure that change would be welcome from players who are used to the game remembering they want to doublefire rather than having to set it every mission. Maybe there needs to be a doublefire flag validity check during mission start?

[the remark this answers is not in the document]
Comment (Goober5000) [reply code/ship/ship.cpp:11934]: Ship flags are cleared between missions, and also when you restart missions.  I don't see how a doublefire flag could carry over under any of those circumstances.  If it does, that's a bug.  In that case, upload a simple test mission and I'll take a look.

An assertion is the correct way to handle those cases.  An assertion is expected to never trigger, but it still performs the check in debug builds just in case.

[the remark this answers is not in the document]
Comment (EatThePath) [reply code/ship/ship.cpp:11934]: Ship flags are cleared yes, but further investigation revealed what I really should have expected in the first place; several things are stored in the player data and restored from it on mission start, the doublefire flag among them. Nothing I can see suggest it's supposed to be tied to a specific mission, just generally remembering the 'cockpit state' as it were of the player's ship as it was at the end of last session.

So, I probably ought to add the flag validity check in or after the function that does that, and then move things around in key control, and finally replace the stuff here for a sanity assert as you suggest. probably a monday project.

**Comment by EatThePath** [comment]
The commit log is noisier than I intended because I apparently don't know how to squash properly, but this should overall be a bit more complete and proper.

File: code/ai/aicode.cpp [hunk 4]
```diff
@@ -5520,7 +5520,10 @@ void set_primary_weapon_linkage(object *objp)
 				if (!sip->flags[Ship::Info_Flags::No_primary_linking] ) {
 					shipp->flags.set(Ship::Ship_Flags::Primary_linked);
 				}
-                shipp->flags.set(Ship::Ship_Flags::Secondary_dual_fire);
```
Comment (Goober5000) [inline code/ai/aicode.cpp:5523]: Code rot has left this section out-of-order and muddled.  Starting with `shipp = &Ships[objp->instance];` and extending until `// AL 2-11-98: If ship has a disarm or disable goal, don't link unless both weapons are` , replace the code with this...

```
	shipp = &Ships[objp->instance];

	if (Num_weapons > (int) (MAX_WEAPONS * 0.75f)) {
		if (shipp->flags[Ship::Ship_Flags::Primary_linked]) {
			nprintf(("AI", "Frame %i, ship %s: Unlinking primaries.\n", Framecount, shipp->ship_name));
			shipp->flags.remove(Ship::Ship_Flags::Primary_linked);
		}
		return;		//	If low on slots, don't link.
	}

	sip = &Ship_info[shipp->ship_info_index];
	aip	= &Ai_info[shipp->ai_index];
	swp = &shipp->weapons;

	shipp->flags.remove(Ship::Ship_Flags::Primary_linked);

	// AL: ensure target is a ship!
	if ( (aip->target_objnum != -1) && (Objects[aip->target_objnum].type == OBJ_SHIP) ) {
		// If trying to destroy a big ship (i.e., not disable/disarm), always unleash all weapons
		if ( Ship_info[Ships[Objects[aip->target_objnum].instance].ship_info_index].is_big_ship() ) {
			if ( aip->targeted_subsys == nullptr ) {
				if (!sip->flags[Ship::Info_Flags::No_primary_linking] ) {
					shipp->flags.set(Ship::Ship_Flags::Primary_linked);
				}
				if (!(The_mission.ai_profile->flags[AI::Profile_Flags::Disable_ai_secondary_doublefire] ||
					Weapon_info[swp->secondary_bank_weapons[swp->current_secondary_bank]].wi_flags[Weapon::Info_Flags::No_doublefire])) {
					shipp->flags.set(Ship::Ship_Flags::Secondary_dual_fire);
				}
				return;
			}
		}
	}

	// AL 2-11-98: If ship has a disarm or disable goal, don't link unless both weapons are
```

The point of this change is to move the `MAX_WEAPONS * 0.75f` check to the beginning of the function, where it belongs (and was originally in the public code release), and to decouple it from the flag check.

**Comment by Goober5000** [comment]
After scrutinizing both the PR and the code, I've concluded that this is quickly becoming a very ugly series of changes and I think it should be vastly simplified.  It should work exactly the same way as the "nolink" primary flag: allow the double-fire flag to be set on a shipwide basis unless there is a shipwide "no-doublefire" flag.  Additionally, only check the no-doublefire flag *on the weapon* when the weapon is actually fired.

By the way, that aicode.cpp.orig flag should not be included in this PR.

**Comment by EatThePath** [comment]
> 
> 
> After scrutinizing both the PR and the code, I've concluded that this is quickly becoming a very ugly series of changes and I think it should be vastly simplified. It should work exactly the same way as the "nolink" primary flag: allow the double-fire flag to be set on a shipwide basis unless there is a shipwide "no-doublefire" flag. Additionally, only check the no-doublefire flag _on the weapon_ when the weapon is actually fired.

Other restructuring aside if you only check the flag when fired, then the player can be shown they are capable of doublefiring when they actually can't, potentially quite frequently for the weapon flag case. In https://github.com/scp-fs2open/fs2open.github.com/pull/2328#discussion_r409838607 you concluded that would be bad. Has that position changed?

> 
> By the way, that aicode.cpp.orig flag should not be included in this PR.

Sorry about that, I'll clean it up. Presumably I overlooked my merge tool making it, or somesuch.

**Comment by Goober5000** [comment]
> Other restructuring aside if you only check the flag when fired, then the player can be shown they are capable of doublefiring when they actually can't, potentially quite frequently for the weapon flag case. In #2328 (comment) you concluded that would be bad. Has that position changed?

What has changed is my understanding of the implementation.  The core concept, "Hiding important feedback from the player is a bad thing," remains the same.  The player should still be able to tell whether double-fire is allowed.

But I realized while looking over the code that you can add a special check to not draw the double-fire indicator if the bank is single-fire only.  The ship overall is still in double-fire mode.  This has the advantage that the double-fire state is preserved if the player switches between weapons with and without the flag.  It also means that the AI will not forget to double-fire if it similarly switches between weapons.

**Comment by Goober5000** [comment]
Implemented in #2347
